receiver/kafkareceiver/kafka_receiver.go

```diff
@@ -0,0 +1,165 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kafkareceiver
+
+import (
+	"context"
+
+	"github.com/Shopify/sarama"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+	"go.uber.org/zap"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/obsreport"
+)
+
+const (
+	transport = "kafka"
+)
+
+// kafkaConsumer uses sarama to consume and handle messages from kafka.
+type kafkaConsumer struct {
+	name              string
+	consumerGroup     sarama.ConsumerGroup
+	nextConsumer      consumer.TraceConsumer
+	topics            []string
+	cancelConsumeLoop context.CancelFunc
+	unmarshaller      unmarshaller
+
+	logger *zap.Logger
+}
+
+var _ component.Receiver = (*kafkaConsumer)(nil)
+
+func newReceiver(config Config, params component.ReceiverCreateParams, nextConsumer consumer.TraceConsumer) (*kafkaConsumer, error) {
+	c := sarama.NewConfig()
+	c.ClientID = config.ClientID
+	c.Metadata.Full = config.Metadata.Full
+	c.Metadata.Retry.Max = config.Metadata.Retry.Max
+	c.Metadata.Retry.Backoff = config.Metadata.Retry.Backoff
+	if config.ProtocolVersion != "" {
+		version, err := sarama.ParseKafkaVersion(config.ProtocolVersion)
+		if err != nil {
+			return nil, err
+		}
+		c.Version = version
+	}
+	client, err := sarama.NewConsumerGroup(config.Brokers, config.GroupID, c)
+	if err != nil {
+		return nil, err
+	}
+	return &kafkaConsumer{
+		name:          config.Name(),
+		consumerGroup: client,
+		topics:        []string{config.Topic},
+		nextConsumer:  nextConsumer,
+		unmarshaller:  &protoUnmarshaller{},
+		logger:        params.Logger,
+	}, nil
+}
+
+func (c *kafkaConsumer) Start(context.Context, component.Host) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	c.cancelConsumeLoop = cancel
+	consumerGroup := &consumerGroupHandler{
+		name:         c.name,
+		logger:       c.logger,
+		unmarshaller: c.unmarshaller,
+		nextConsumer: c.nextConsumer,
+		ready:        make(chan bool),
+	}
+	go c.consumeLoop(ctx, consumerGroup)
+	<-consumerGroup.ready
+	return nil
+}
+
+func (c *kafkaConsumer) consumeLoop(ctx context.Context, handler sarama.ConsumerGroupHandler) error {
+	for {
+		// `Consume` should be called inside an infinite loop, when a
+		// server-side rebalance happens, the consumer session will need to be
+		// recreated to get the new claims
+		if err := c.consumerGroup.Consume(ctx, c.topics, handler); err != nil {
```

Comment:
Is it not a problem that most likely this will block forever if no messages are coming, so never gets closed?

Reply:
When no messages are coming the call passes and the handler is ready to receive messages. 

When the close on the receiver is called it closes the consumer group correctly and interrupts the infinite loop via the `cancelConsumeLoop context.CancelFunc` (the  "cancel" question you had before).